### PR TITLE
refactor! [BREAKING] Switching to Project paradigm

### PR DESCRIPTION
**Pull request by JigarJoshi**
See https://github.com/tigrisdata/tigris-api/pull/166 & https://github.com/tigrisdata/tigris-api/pull/159

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -28,7 +28,6 @@ type Driver struct {
 	Token        string        `json:"token,omitempty"`
 	URL          string        `json:"url,omitempty"`
 	Protocol     string        `json:"protocol,omitempty"`
-	Project      string        `json:"project,omitempty"`
```
Comment (ovaistariq) [inline config/config.go:31]: What is the reason for this change?

File: tigris/client.go [hunk 2]
```diff
@@ -31,7 +31,6 @@ type Config struct {
 	Token        string      `json:"token,omitempty"`
 	URL          string      `json:"url,omitempty"`
 	Protocol     string      `json:"protocol,omitempty"`
-	Project      string      `json:"project,omitempty"`
```
Comment (ovaistariq) [inline tigris/client.go:34]: What is the reason for removing this?

Comment (efirs) [reply tigris/client.go:34]: +1

File: config/config.go (hunk 1 above)
Comment (JigarJoshi) [reply config/config.go:31]: Because driver is not associated with project anymore. Driver can create new project and it takes project name as argument for appropriate methods.

File: tigris/client.go (hunk 2 above)
Comment (JigarJoshi) [reply tigris/client.go:34]: same as above.

File: config/config.go (hunk 1 above)
Comment (ovaistariq) [reply config/config.go:31]: The client object should still be associated with a project I think because except for a few operations all the operations are within the scope of a single project. Can project creation/deletion functions be outside the close object?
From a user perspective, a regular user would always pass project when instantiating the client object so that all operations are properly scoped. Which is why it is better to have this in the config

Comment (JigarJoshi) [reply config/config.go:31]: ok - higher level client (Client interface from tigris package) is now associated with project.

Comment (ovaistariq) [reply config/config.go:31]: Would this still be in the config?

File: driver/driver.go [hunk 3]
```diff
@@ -376,13 +402,6 @@ func initConfig(lCfg *config.Driver) (*config.Driver, error) {
 		cfg.URL = DefaultURL
 	}
 
-	if cfg.Project == "" {
```
Comment (ovaistariq) [inline driver/driver.go:379]: Do you want to remove the rest of the environment handling from here as well?

File: tigris/example_test.go [hunk 4]
```diff
@@ -63,7 +63,7 @@ func ExampleOpenDatabase() {
 	// Connects to the Tigris server. Creates or opens database "db1".
 	// Creates or migrate &Coll1{}. Returns a "db" object, which provides
 	// access to the collections of the database, Coll1 in this example.
-	db, err := OpenDatabase(ctx, &Config{Project: "db1"}, "db1", &Coll1{})
+	db, err := OpenDatabase(ctx, &Config{}, "db1", &Coll1{})
```
Comment (efirs) [inline tigris/example_test.go:66]: We need to move db1 inside tigris.Config?

File: tigris/client.go [hunk 5]
```diff
@@ -65,6 +66,13 @@ func NewClient(ctx context.Context, cfg ...*Config) (*Client, error) {
 		pCfg = *cfg[0]
 	}
 
+	if pCfg.Project == "" {
```
Comment (ovaistariq) [inline tigris/client.go:69]: IMO this should happen where the rest of the environment parsing is happening otherwise it is not consistent

File: tigris/example_test.go (hunk 4 above)
Comment (JigarJoshi) [reply tigris/example_test.go]: yes this was left. thanks!

File: driver/driver.go (hunk 3 above)
Comment (efirs) [reply driver/driver.go:379]: We should not.
Direct user or the driver: CLI, Ferret, depend on this functionality.

The only variable we need to move up to the high level client is "project", because driver is not bound to the project and require project to be specified explicitly.

File: tigris/client.go (hunk 5 above)
Comment (efirs) [reply tigris/client.go:69]: see above

**Review by efirs — APPROVED** [review]

File: driver/driver.go (hunk 3 above)
Comment (ovaistariq) [reply driver/driver.go:379]: That doesn't make the code clean then as environment handling is partially happening in two different places.
